### PR TITLE
[WS1][kernels] Batch-invariant attention (standard softmax--cuda version) 

### LINKED ISSUE #147: [WS1][kernels] Batch-invariant attention (standard softmax)
Part of WS1 — Full Batch-Invariant Forward Chain (epic: #<WS1 tracking issue>)

## Why

Flash-style attention parallelizes over the KV dimension and merges partial results, and the number of KV splits depends on sequence length and batch — so the online-softmax accumulation order changes with batch configuration, breaking invariance. Because attention mixes information across positions, even tiny drift here spreads to every later token.

## Scope

Provide a batch-invariant standard-softmax attention for the forward chain.

- Start from a deterministic masked-softmax + logsumexp (LSE) kernel with a fixed max / sum-exp order, then extend to the full attention output (P x V).
- Avoid split-KV / variable KV-partitioning; use a single-SM whole-sequence reduction, or a dual-kernel design with an identical, fixed accumulation order regardless of split count.
- Causal mask + padding mask support; variable-length sequence coverage.
- Keep prefill and decode on a consistent reduction path (coordinate with the KV-cache consistency issue); expose the hooks that issue needs.
- Accumulate softmax statistics in FP32.
- Validate against the #108 harness across batch=1/N, chunked-prefill on/off, padding layouts.

## Out of scope

- Non-standard attention variants (linear / sliding-window / MoE routing) — standard softmax only for this gate.
- Multi-GPU / sequence-parallel attention (WS2).
- FP8 attention.
- Full FlashAttention replacement / maximum throughput in the first deterministic baseline.

## Acceptance criteria

- The same query / sequence produces bitwise-identical (or within #108 tolerance) attention output regardless of batch size, batch position, and chunked-prefill on/off.
- No split-KV path whose split count varies with batch / seq config; the accumulation order is pinned and documented.
- Softmax + LSE outputs are invariant across the sweep; tests cover both fixed-length and variable-length sequences.
- Decode-stage output matches the prefill path for the same effective context (handoff with the KV-cache issue).
- The attention backward path passes the shared gradient-invariance check from the WS1 backward-consistency issue.

## Notes

- Tightly coupled with the KV-cache path-consistency issue — agree on the shared reduction contract early.
- Existing attention kernels (#65 prefix-shared fused attention, #41 Triton FlashAttention fallback) are reference / audit starting points, not the invariant path itself.
- Single-SM whole-sequence is simplest to prove correct but may cost throughput on long sequences; note the tradeoff if choosing the dual-kernel route.

## Planned PRs

- [ ] Deterministic masked-softmax + LSE kernel (fixed max / sum-exp order)
- [ ] Causal + padding mask tests; variable-length coverage
- [ ] Deterministic P x V aggregation (single-SM or fixed-order dual-kernel)
- [ ] Prefill/decode handoff hooks (coordinate with the KV-cache issue)
- [ ] Wire through the #108 harness sweep; benchmark + tradeoff note

### PR DESCRIPTION
**Pull request by maxiaosong1124**
## Summary

  Adds a deterministic CUDA standard-softmax attention implementation for #147.

  The implementation provides batch-invariant forward and backward paths with a fixed accumulation order. It supports causal masking, key-padding masks, GQA/MQA, variable sequence lengths, chunked prefill, decode, and KV-cache handoff.

  The CUDA path matches the existing FP32 golden reference NativeAttentionOp within the shared #108 tolerance contract. Outputs and gradients are bitwise invariant when the same sequence is evaluated at different batch sizes, batch positions, or prefill chunk sizes.

  Closes #147.

  ## Implementation

  The forward path is decomposed into three deterministic CUDA kernels:

  1. QK: computes FP32 attention scores with a fixed d = 0 ... D-1 accumulation order.
  2. Masked softmax + LSE: one CTA per attention row with a fixed 256-thread shared-memory tree reduction.
  3. PV: computes the output with a fixed k = 0 ... Skv-1 FP32 accumulation order.

  The backward path implements deterministic:

  - dP
  - softmax backward
  - dQ
  - dK
  - dV

  For GQA, dK and dV accumulate query heads and query positions in a fixed order. There are no cross-CTA atomics, split-KV paths, or launch-order-dependent reductions.

   Path               Status
  ━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   CUDA forward       Deterministic QK → masked softmax/LSE → PV
  ─────────────────  ───────────────────────────────────────────────────────
   CUDA backward      Deterministic dP → softmax backward → dQ/dK/dV
  ─────────────────  ───────────────────────────────────────────────────────
   Accumulation       FP32 intermediates and statistics
  ─────────────────  ───────────────────────────────────────────────────────
   Dtypes             FP16 and BF16
  ─────────────────  ───────────────────────────────────────────────────────
   Head dimension     D=128
  ─────────────────  ───────────────────────────────────────────────────────
   Attention modes    Causal and non-causal
  ─────────────────  ───────────────────────────────────────────────────────
   Head layouts       MHA, MQA, and GQA
  ─────────────────  ───────────────────────────────────────────────────────
   Padding            key_padding_mask support, including fully masked rows
  ─────────────────  ───────────────────────────────────────────────────────
   Prefill/decode     Shared kernel and causal-offset contract
  ─────────────────  ───────────────────────────────────────────────────────
   Registry           CUDA_DETERMINISTIC_ATTENTION with native fallback
  ─────────────────  ───────────────────────────────────────────────────────
   Harness            Integrated with the #108 operator suite
  ─────────────────  ───────────────────────────────────────────────────────
   Benchmark          Qwen3-8B representative prefill/decode shapes

  ## Fixed Reduction Contract

  ### QK

  Each score has exactly one writer. The head dimension is accumulated in ascending order:

  for d = 0 .. D-1:
      score += float(q[d]) * float(k[d])

  ### Softmax and LSE

  Each (batch, query_head, query_position) row is processed by one CTA.

  Max and sum-exp use a fixed power-of-two shared-memory tree reduction. The reduction topology does not depend on batch size, sequence length, SM count, or runtime split count.

  ### PV

  Each output element has exactly one writer:

  for k = 0 .. Skv-1:
      out += probability[k] * float(value[k])

  ### GQA backward

  Each dK and dV element is accumulated by one thread:

  for local_head = 0 .. group_size-1:
      for query_position = 0 .. Sq-1:
          grad += ...

  No split-KV or cross-CTA atomic accumulation is used.

  ## Prefill / Decode / KV-cache Contract

  All modes use the same deterministic attention kernels:

  - Prefill: Sq == Skv
  - Chunked prefill: Sq = chunk_size, Skv = past + chunk_size
  - Decode: Sq = 1 or a small number of new tokens
  - KV-cache handoff: concatenate cached and new K/V, then call the same attention path

  The causal condition is:

  key_index <= Skv - Sq + query_index

  This keeps full prefill, chunked prefill, and decode on the same effective reduction path.

  The Python API exposes:

  - forward(...) for the normal differentiable path
  - forward_with_lse(...) for output/LSE validation and KV-cache integration

  ## Acceptance Criteria

   Requirement                            Result
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   Batch-size invariance                  ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Batch-position invariance              ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Batch chunking invariance              ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Chunked-prefill on/off                 ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Chunked-prefill with padding           ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Softmax and LSE invariance             ✅ Bitwise for matching physical shapes
  ─────────────────────────────────────  ─────────────────────────────────────────
   Fixed and variable sequence lengths    ✅
  ─────────────────────────────────────  ─────────────────────────────────────────
   Causal and padding masks               ✅
  ─────────────────────────────────────  ─────────────────────────────────────────
   Prefill/decode handoff                 ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   KV-cache concatenation handoff         ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Backward correctness                   ✅ Within #108 tolerance
  ─────────────────────────────────────  ─────────────────────────────────────────
   Gradient batch invariance              ✅ Bitwise
  ─────────────────────────────────────  ─────────────────────────────────────────
   Fixed accumulation order               ✅ Documented and tested
  ─────────────────────────────────────  ─────────────────────────────────────────
   No variable split-KV path              ✅

  A valid-only sequence and a physically padded sequence use different reduction widths. That comparison is therefore checked within the shared accuracy tolerance rather than bitwise. Batch placement and chunking with the same effective physical shape remain bitwise invariant.

  ## Validation Environment

   Item                  Value
  ━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
   GPU                   NVIDIA H20
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   Compute capability    9.0
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   GPU memory            97,871 MiB
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   Driver                580.65.06
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   PyTorch               2.8.0+cu128
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   CUDA runtime          12.8
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   Python                3.12.3
  ────────────────────  ───────────────────────────────────────────────────────────────────────────
   Extension             Editable CUDA build with deterministic attention forward/backward symbols

  ## Correctness / Tests

  ### Issue #147 CUDA suite

  python -m pytest tests/test_deterministic_attention_cuda.py -q -ra

  Result:

  609 passed in 3.38s

  Coverage includes:

  - FP16/BF16
  - MHA/MQA/GQA
  - causal and non-causal attention
  - fixed and non-power-of-two sequence lengths
  - padding and fully masked rows
  - custom/default/zero scale
  - output and LSE correctness
  - batch slice, permutation, and chunk invariance
  - sequence-dimension chunked prefill
  - prefill/decode and KV-cache handoff
  - forward and backward #108 harness integration
  - FP64 gradient reference comparison
  - gradient batch invariance
  - fixed GQA dK/dV accumulation order

  ### Attention integration tests

  python -m pytest \
      tests/test_attention.py \
      tests/test_operator_inputs.py \
      tests/test_kernel_registry.py \
      -q -ra

  Result:

  50 passed in 2.54s

  ### Full repository suite

  python -m pytest -q -ra

  Result:

  1165 passed, 102 skipped in 19.42s

  The skipped tests are environment-specific FlashAttention, ROCm, CI-only extension enforcement, and optional SM90 fused-logp tests. No deterministic-attention test was skipped.

  ### Pre-commit

  pre-commit run --all-files

  All hooks pass:

  - trailing whitespace
  - end-of-file fixer
  - YAML validation
  - large-file check
  - Black
  - isort
  - flake8

  ## Benchmark

  Command:

  python benchmarks/benchmark_deterministic_attention.py \
      --dtype bf16 \
      --warmup 1 \
      --iters 2

  Environment: NVIDIA H20. These numbers are smoke-test measurements rather than a formal performance study.

   Shape                 Latency (ms)    Peak memory (MiB)    Scores (MiB)
  ━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━
   decode-128                   0.057                  0.5             0.0
  ────────────────────  ──────────────  ───────────────────  ──────────────
   decode-1k                    0.121                  4.1             0.1
  ────────────────────  ──────────────  ───────────────────  ──────────────
   prefill-128                  0.344                  4.5             2.0
  ────────────────────  ──────────────  ───────────────────  ──────────────
   prefill-512                  4.821                 42.1            32.0
  ────────────────────  ──────────────  ───────────────────  ──────────────
   prefill-1k                  19.150                148.1           128.0
  ────────────────────  ──────────────  ───────────────────  ──────────────
   batch4-prefill-128           1.287                 18.1             8.0
  ────────────────────  ──────────────  ───────────────────  ──────────────
   batch8-prefill-64            0.687                 14.1             4.0

  ## Memory / Performance Tradeoff

  The first deterministic baseline materializes full FP32 tensors:

  scores: [B, Hq, Sq, Skv]
  P:      [B, Hq, Sq, Skv]

  This gives a simple, auditable fixed reduction order but has quadratic sequence-length memory usage and is not intended to replace FlashAttention for maximum throughput.

  A future implementation may use a fixed-order dual-kernel design, provided it preserves the same batch-invariance contract.

  ## Limitations

  - CUDA only
  - FP16 and BF16 inputs
  - Head dimension currently fixed to 128
  - Standard softmax attention only
  - Full FP32 score/probability materialization
  - No FP8, sliding-window, sequence-parallel, or multi-GPU attention

## Summary by CodeRabbit

* **New Features**
  * Added a deterministic CUDA standard-softmax attention backend with causal masking, key padding mask support, and grouped-query attention.
  * Updated attention dispatch to prefer the deterministic CUDA backend when available.
  * Added an optional interface to return log-sum-exp (LSE) for verification/debugging.

* **Documentation**
  * Expanded attention operator documentation with deterministic guarantees, backend constraints/limitations, and memory tradeoffs.

* **Tests**
  * Added end-to-end deterministic attention forward/backward coverage (padding, chunking, cache handoff, and determinism checks).
  * Extended accuracy tolerance contracts for the attention operator.

* **Chores**
  * Added a CUDA benchmarking script for representative attention shapes.

### CODE REVIEW

File: tests/test_deterministic_attention_cuda.py [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+        name=name,
+        op_class="attention",
+        dtype=dtype,
+        inputs=inputs,
+        gold_fn=gold.forward_fp32,
+        grad_input_names=("q", "k", "v"),
+    )
+
+
+HARNESS_CASES = [
```
Comment (Flink-ddd) [inline tests/test_deterministic_attention_cuda.py]: Defer the creation of CUDA tensors in HARNESS_CASES until runtime to prevent pytest collection failures on CPU-only CI nodes.

File: csrc/cuda/attention/deterministic_attention.cu [hunk 2]
```diff
@@ -0,0 +1,651 @@
...
+  }
+
+  const int64_t B = q_contig.size(0);
+  const int64_t Hq = q_contig.size(1);
+  const int64_t Sq = q_contig.size(2);
+  const int64_t D = q_contig.size(3);
+  const int64_t Hkv = k_contig.size(1);
+  const int64_t Skv = k_contig.size(2);
+
+  auto stream = at::cuda::getCurrentCUDAStream();
```
Comment (Flink-ddd) [inline csrc/cuda/attention/deterministic_attention.cu:286]: Insert at::cuda::OptionalCUDAGuard before acquiring the CUDA stream to ensure kernels launch on the correct tensor device and prevent illegal memory accesses.

File: csrc/cuda/attention/deterministic_attention.cu [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+  auto P_c = P.contiguous();
+
+  const int64_t B = q_c.size(0);
+  const int64_t Hq = q_c.size(1);
+  const int64_t Sq = q_c.size(2);
+  const int64_t D = q_c.size(3);
+  const int64_t Hkv = k_c.size(1);
+  const int64_t Skv = k_c.size(2);
+
+  auto stream = at::cuda::getCurrentCUDAStream();
```
Comment (Flink-ddd) [inline csrc/cuda/attention/deterministic_attention.cu:567]: same

File: rl_engine/kernels/ops/cuda/attention/deterministic_attn.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        q_c = q.contiguous()
+        k_c = k.contiguous()
+        v_c = v.contiguous()
+        mask_c = key_padding_mask.contiguous() if key_padding_mask is not None else None
+
+        results = _C.deterministic_attention_forward(q_c, k_c, v_c, causal, float(scale), mask_c)
+        out, lse, P = results[0], results[1], results[2]
+
+        ctx.save_for_backward(q_c, k_c, v_c, P)
```
Comment (Flink-ddd) [inline rl_engine/kernels/ops/cuda/attention/deterministic_attn.py]: Save all required tensors using ctx.save_for_backward to prevent memory leaks, and add the @once_differentiable decorator to the backward function to explicitly fail unsupported higher-order gradient requests.

**Review by Flink-ddd — CHANGES_REQUESTED** [review]
Left a few comments, otherwise LGTM.

**Comment by Flink-ddd** [comment]
please resolve the code conflicts first, Thank you.

**Comment by maxiaosong1124** [comment]
> please resolve the code conflicts first, Thank you.

@Flink-ddd, thank you for the review.

  I’ve resolved the merge conflicts and addressed all four comments in commit 040acd9:

  - Deferred CUDA tensor creation until test execution, so collection works on CPU-only nodes.
  - Added OptionalCUDAGuard to both forward and backward CUDA wrappers.
  - Saved the key-padding mask with ctx.save_for_backward.
  - Added @once_differentiable to explicitly reject unsupported higher-order gradients.

  I also ran Black, Flake8, and the CPU-only collection checks successfully. The updated changes have been pushed. Could you please take another look when convenient?

**Review by Flink-ddd — APPROVED** [review]
LGTM now, Thank you for update. also cc @KJLdefeated @inaniloquentee @frank-2077 PTAL.

**Review by KJLdefeated — APPROVED** [review]
LGTM!
